File: coursera/kotlin-for-java-developers/week4/src/main/java/properties/Properties.java
package properties;

import org.jetbrains.annotations.NotNull;

public class Properties {

    public static void main(String[] args) {
        ContactKotlin contactKotlin = new ContactKotlin("Luke", "London");
        System.out.println("contact: " + contactKotlin.getName() + ", lives in: " + contactKotlin.getAddress());

        ContactJava contactJava = new ContactJava("Luke", "London");
        System.out.println("contact: " + contactJava.getName() + ", lives in: " + contactJava.getAddress());

        PersonKotlin personKotlin = new PersonKotlin("Luke", 29);
        System.out.println("person: " + personKotlin.getName() + ", is: " + personKotlin.getAge());

        PersonJava personJava = new PersonJava("Luke", 29);
        System.out.println("person: " + personJava.getName() + ", is: " + personJava.getAge());

    }

    static final class ContactJava {
        @NotNull
        private final String name;
        @NotNull
        private final String address;

        public ContactJava(@NotNull String name, @NotNull String address) {
            this.name = name;
            this.address = address;
        }

        @NotNull
        public final String getName() {
            return name;
        }

        @NotNull
        public final String getAddress() {
            return address;
        }
    }

    static final class PersonJava {
        @NotNull
        private final String name;
        private int age;

        public PersonJava(@NotNull String name, int age) {
            this.name = name;
            this.age = age;
        }

        @NotNull
        public final String getName() {
            return name;
        }

        public final int getAge() {
            return age;
        }

        public final void setAge(int age) {
            this.age = age;
        }
    }

}

File: coursera/kotlin-for-java-developers/week4/src/main/kotlin/properties/Properties.kt
package properties

import properties.Properties.ContactJava
import properties.Properties.PersonJava

fun main() {
    val contactKotlin = ContactKotlin("Luke", "London")
    println("Contact: ${contactKotlin.name}, lives in: ${contactKotlin.address}")

    val contactJava = ContactJava("Luke", "London")
    println("Contact: ${contactJava.name}, lives in: ${contactJava.address}")

    val personKotlin = PersonKotlin("Luke", 29)
    println("Person: ${personKotlin.name}, is: ${personKotlin.age}")

    val personJava = PersonJava("Luke", 29)
    println("Person: ${personJava.name}, is: ${personJava.age}")

    val rectangle = Rectangle(2, 3)
    println(rectangle.isSquare)

    println("$foo1 $foo1 $foo2 $foo2")

    val stateLogger = StateLogger()
    println(stateLogger.state)
    stateLogger.state = true
    println(stateLogger.state)

    val lengthCounter = LengthCounter()
    println(lengthCounter.counter)
    lengthCounter.addWord("Luke")
    println(lengthCounter.counter)
}

class ContactKotlin(val name: String, val address: String)

class PersonKotlin(val name: String, var age: Int)

class Rectangle(val height: Int, var width: Int) {
    val isSquare: Boolean
        get() {
            return height == width
        }
}

val foo1 = run {
    println("Calculating the answer...")
    42
}

val foo2: Int
    get() {
        println("Calculating the answer...")
        return 42
    }

class StateLogger {
    var state = false
        set(value) {
            println("state had changed: $field -> $value")
            field = value
        }
}

class LengthCounter {
    var counter: Int = 0
        private set

    fun addWord(word: String) {
        counter += word.length
    }
}
